### PR TITLE
Implement the build_dir abstraction

### PR DESCRIPTION
**Pull request by tkdchen**
* CLOUDBLD-7977

Signed-off-by: Chenxiong Qi <cqi@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- n/a JSON/YAML configuration changes are updated in the relevant schema
- n/a Changes to metadata also update the documentation for the metadata
- n/a Pull request has a link to an osbs-docs PR for user documentation updates
- n/a New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/dirs.py
Comment (chmeliik) [inline atomic_reactor/dirs.py:153]: Maybe sort the platforms aleady in `__init__`? IMO that would also make the `for_each()` and `for_all_copy()` methods slightly more predictable (though the order still shouldn't be relied on)

File: atomic_reactor/dirs.py
Comment (chmeliik) [inline atomic_reactor/dirs.py:192]: I'd say this should return the files for just one build directory - they are guaranteed to be identical anyway

File: atomic_reactor/dirs.py
Comment (chmeliik) [inline atomic_reactor/dirs.py:205]: Maybe the path should be treated as relative to the BuildDir in this case. I think the callable should be allowed to return relative paths

File: atomic_reactor/dirs.py
Comment (chmeliik) [inline atomic_reactor/dirs.py:207]: I think we will have valid use cases for directories. For example the Cachito plugin, it unpacks a tarball into the build dir and I think it's OK to just return the path to the unpacked dir, the plugin shouldn't have to list all regular files

File: atomic_reactor/dirs.py
Comment (chmeliik) [inline atomic_reactor/dirs.py:212]: Big +1 for this, but I think the check could be a bit more robust:
```python
build_dir in filepath.parents
# or
try:
    filepath.relative_to(build_dir)
except ValueError:
    ...
```
The current code may miss some edge cases like
```python
"/foobar/some-file".startswith("/foo")
```

File: atomic_reactor/dirs.py
Comment (chmeliik) [inline atomic_reactor/dirs.py:223]:
```suggestion
                dest = self.path / platform.value / src_file.relative_to(build_dir)
                dest.parent.mkdir(parents=True, exist_ok=True)
```
Would this be simpler? I mainly wanted to avoid the string replacement, it could cause problems if the src_file path contains platform names in unexpected places

File: atomic_reactor/dirs.py
Comment (chmeliik) [inline atomic_reactor/dirs.py:30]: I can't decide how I feel about using the Platform enum vs. a plain string. On the one hand, I do think it's good to limit the valid platforms, but on the other, a plain string would usualy be nicer to work with. I'm also not sure if we should hardcode the valid platforms, since technically they are configurable in the reactor config map.

File: atomic_reactor/dirs.py
Comment (chmeliik) [reply atomic_reactor/dirs.py:151]: > I think we lack information whether the order of passed platforms matters for the caller

In my opinion, we should never have code that relies on processing the platforms in a specific order

> For what cases of this predictability?

We will have at least two different ways to create the RootBuildDir - in the prebuild task, it will be done by the `check_and_set_platforms` plugin, in all other tasks after that, it will be done differently. If the order is up to the caller, then we would have to make sure every initialization passes the platforms in the same order, otherwise different tasks could process platforms in different order (which wouldn't matter that much, since IMO the order shouldn't matter, but it would still be nice to ensure consistency).

File: atomic_reactor/dirs.py
Comment (chmeliik) [reply atomic_reactor/dirs.py:30]: I meant that if we use a string for the platform name, then we have infinitely many possible values, the `Platform` enum has just 4 (and that's a good thing). But I don't know if we can really enumerate all the possible platforms in the upstream code, our instance of OSBS only handles those 4 but I'm not sure if those are the only options, even our docs mention another one: https://osbs.readthedocs.io/en/osbs_ocp3/users.html?highlight=armhfp#image-configuration

File: atomic_reactor/dirs.py
Comment (chmeliik) [reply atomic_reactor/dirs.py:213]: Ooh, I didn't know about [is_relative_to](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to), nice!

But it's new in 3.9, so we can't use it yet :cry:

Comment (tkdchen) [reply atomic_reactor/dirs.py:213]: Oh, I didn't notice that. I'll change it to `relative_to`. :sweat_smile:

File: atomic_reactor/dirs.py
Comment (chmeliik) [inline atomic_reactor/dirs.py:208]: Probably shouldn't be allowed

File: tests/test_dirs.py
Comment (chmeliik) [inline tests/test_dirs.py:27]: You can use [tmp_path](https://docs.pytest.org/en/6.2.x/tmpdir.html#the-tmp-path-fixture) instead of `tmpdir` (completely optional, I'm fine with keeping it this way)

File: tests/test_dirs.py
Comment (chmeliik) [inline tests/test_dirs.py:80]: Oh, I didn't realize that `NewType` requires you to wrap the value in the new type. Would a type alias be enough? https://docs.python.org/3/library/typing.html#type-aliases

**Review by chmeliik — COMMENTED** [review]
A few more minor things, overall looks very nice!

@tkdchen if I remember corretly, you mentioned you would update a few plugins to use this, are you planning to do that in a follow-up PR?

File: tests/test_dirs.py
Comment (ben-alkov) [inline tests/test_dirs.py:10]: 🎉

File: tests/test_dirs.py
Comment (tkdchen) [reply tests/test_dirs.py:80]: Cool. Type alias is better in this case. :+1:

File: atomic_reactor/dirs.py
Comment (tkdchen) [reply atomic_reactor/dirs.py]: Done. Updated to use `relative_to`.

File: atomic_reactor/dirs.py
Comment (tkdchen) [reply atomic_reactor/dirs.py]: Platform enum is removed and the platform is passed as string now.

File: atomic_reactor/dirs.py
Comment (tkdchen) [reply atomic_reactor/dirs.py]: These lines are removed from the docstring.

File: atomic_reactor/dirs.py
Comment (tkdchen) [reply atomic_reactor/dirs.py]: Now, treat it as an error.

**Comment by tkdchen** [comment]
@chmeliik 

> if I remember corretly, you mentioned you would update a few plugins to use this, are you planning to do that in a follow-up PR?

Yes. That is for making an example by using this new API. I chose plugin flatpak_create_dockerfile to update. More time is needed. So I think once this PR is good enough, it can be merged firstly and I'll open another PR for the plugin update.

**Review by chmeliik — COMMENTED** [review]
LGTM :+1:

The only question left for me is if we can ignore the `.git/` dir or not when copying

File: atomic_reactor/dirs.py
Comment (MartinBasti) [reply atomic_reactor/dirs.py]: Git should stay, people may use git ref during their builds in mutlistage builds, or weherever.

**Comment by MartinBasti** [comment]
> The only question left for me is if we can ignore the `.git/` dir or not when copying

No, keep it

**Comment by MartinBasti** [comment]
LGTM (I just read it quickly, so I'm not sure enough to give full ACK)

**Comment by tkdchen** [comment]
Hi, this PR needs approvals for merge. :)

**Review by chmeliik — APPROVED** [review]
LGTM :+1:

**Review by mkosiarc — APPROVED** [review]

File: atomic_reactor/dirs.py
Comment (lkolacek) [inline atomic_reactor/dirs.py:38]: double path

**Review by lkolacek — APPROVED** [review]

File: atomic_reactor/dirs.py
Comment (tkdchen) [reply atomic_reactor/dirs.py]: @lkolacek Thank you. :)

**Pull request merged**
